src/platform_impl/linux/x11/window.rs

```diff
@@ -214,30 +214,44 @@ impl UnownedWindow {
             None => xconn.default_screen_index() as c_int,
         };
 
+        // An iterator over all of the visuals combined with their depths.
+        let mut all_visuals = xconn
+            .xcb_connection()
+            .setup()
+            .roots
+            .iter()
+            .flat_map(|root| &root.allowed_depths)
+            .flat_map(|depth| {
+                depth
+                    .visuals
+                    .iter()
+                    .map(move |visual| (visual, depth.depth))
+            });
+
         // creating
-        let (visual, depth, require_colormap) = match pl_attribs.visual_infos {
-            Some(vi) => (vi.visualid as _, vi.depth as _, false),
+        let (visualtype, depth, require_colormap) = match pl_attribs.visual_id {
+            Some(vi) => {
+                // Find this specific visual.
+                let (visualtype, depth) = all_visuals
+                    .find(|(visual, _)| visual.visual_id == vi)
+                    .ok_or_else(|| os_error!(OsError::XError(X11Error::NoSuchVisual(vi).into())))?;
+
+                (Some(visualtype), depth, true)
+            }
             None if window_attrs.transparent => {
                 // Find a suitable visual, true color with 32 bits of depth.
-                xconn.xcb_connection().setup().roots
-                    .iter()
-                    .flat_map(|root| &root.allowed_depths)
-                    .filter(|depth| depth.depth == 32)
-                    .flat_map(|depth| depth.visuals.iter().map(move |visual| (visual, depth.depth)))
+                all_visuals
                     .find(|(visual, _)| {
                         visual.class == xproto::VisualClass::TRUE_COLOR
                     })
                     .map_or_else(|| {
                         debug!("Could not set transparency, because XMatchVisualInfo returned zero for the required parameters");
-                        (x11rb::COPY_FROM_PARENT as _, x11rb::COPY_FROM_PARENT as _, false)
-                    }, |(visual, depth)| (visual.visual_id, depth, true))
+                        (None as _, x11rb::COPY_FROM_PARENT as _, false)
+                    }, |(visual, depth)| (Some(visual), depth, true))
             }
-            _ => (
-                x11rb::COPY_FROM_PARENT as _,
-                x11rb::COPY_FROM_PARENT as _,
-                false,
-            ),
+            _ => (None, x11rb::COPY_FROM_PARENT as _, false),
         };
+        let visual = visualtype.map_or(x11rb::COPY_FROM_PARENT, |v| v.visual_id);
```

Comment:
Move this to the place it's used, you can inline it inside the `match`.

Reply:
It's also used in `create_window`.

Reply:
This is impressive how messy this all is, given that colormap code is also using that visual stuff.

Reply:
Yes, this backend need a serious refactoring.